workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/AndroidDialogBounds.kt

```diff
@@ -0,0 +1,78 @@
+package com.squareup.workflow1.ui.container
+
+import android.app.Dialog
+import android.graphics.Rect
+import android.view.Gravity
+import android.view.View
+import android.view.Window
+import com.squareup.workflow1.ui.ViewEnvironment
+import com.squareup.workflow1.ui.WorkflowUiExperimentalApi
+import com.squareup.workflow1.ui.environment
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+
+/**
+ * Updates the size of the [Window] of the receiving [Dialog].
+ * [bounds] is expected to be in global display coordinates,
+ * e.g. as returned from [View.getGlobalVisibleRect].
+ *
+ * @see ModalScreenOverlayDialogFactory.updateBounds
+ */
+@WorkflowUiExperimentalApi
+public fun Dialog.setBounds(bounds: Rect) {
+  window?.let {
+    it.attributes = it.attributes.apply {
+      gravity = Gravity.TOP or Gravity.START
+      width = bounds.width()
+      height = bounds.height()
+      x = bounds.left
+      y = bounds.top
+    }
+  }
+}
+
+@WorkflowUiExperimentalApi
+internal fun <D : Dialog> D.maintainBounds(
+  view: View,
+  onBoundsChange: (D, Rect) -> Unit
+) {
+  maintainBounds(view.environment!!, onBoundsChange)
+}
+
+@WorkflowUiExperimentalApi
+internal fun <D : Dialog> D.maintainBounds(
+  environment: ViewEnvironment,
+  onBoundsChange: (D, Rect) -> Unit
+) {
+  maintainBounds(environment[ModalArea].bounds, onBoundsChange)
+}
+
+@WorkflowUiExperimentalApi
+internal fun <D : Dialog> D.maintainBounds(
+  bounds: StateFlow<Rect>,
+  onBoundsChange: (D, Rect) -> Unit
+) {
+  val window = requireNotNull(window) { "Dialog must be attached to a window." }
+  window.callback = object : Window.Callback by window.callback {
+    var scope: CoroutineScope? = null
+
+    override fun onAttachedToWindow() {
+      scope = CoroutineScope(Dispatchers.Main.immediate).also {
+        bounds.onEach { b -> onBoundsChange(this@maintainBounds, b) }
+          .launchIn(it)
+      }
+    }
+
+    override fun onDetachedFromWindow() {
+      scope?.cancel()
+      scope = null
+    }
+  }
```

Comment:
👍🏻 .

Other than the injected immediate main dispatcher - but I see thats covered by #600 .

If we want debugging help we could add a name into the context for the scope.

```suggestion
  window.callback = object : Window.Callback by window.callback {
    var scope: CoroutineScope? = null

    override fun onAttachedToWindow() {
      val id = window.decorView.id
      scope = CoroutineScope(Dispatchers.Main.immediate + CoroutineName(id)).also {
        bounds.onEach { b -> onBoundsChange(this@maintainBounds, b) }
          .launchIn(it)
      }
    }

    override fun onDetachedFromWindow() {
      scope?.cancel()
      scope = null
    }
  }
```

Reply:
s/classname/toString()/, 10x more useful. I was shying away from that since it can be big, and that burned us in `workflowAction {}`. But I think that's okay for infrequent events like creating a dialog.

Reply:
Meh, I'm over engineering this.

Gonna merge as is and add a note to #600 to consider useful scope name at the same time.